README.md
# srv

minimalist http(s) server and file browser.

<img src="https://github.com/joshuarli/srv/blob/master/.github/screenshot.png?raw=true" width="200">


## download

static executables for some platforms can be found [here](https://github.com/joshuarli/srv/releases).


## usage

Simply `srv`. Defaults are `-p 8000 -b 127.0.0.1 -d .`


## usage: TLS

TLS and HTTP/2 are enabled if you pass `-c certfile -k keyfile`.

to make self-signed certs:

    openssl req -nodes -new -x509 -keyout key.pem -out cert.pem -subj "/"

or better, locally trusted certs with [mkcert](https://github.com/FiloSottile/mkcert):

    mkcert -install
    mkcert -key-file key.pem -cert-file cert.pem -ecdsa 127.0.0.1


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.6] - Unreleased

Nothing planned at the moment.


## [0.5] - Feb 14 2021
### Fixed
- Correctness fixes with serving more complicated filenames (links needed to be path escaped, and request uris path unescaped).

## [0.4] - 2020-10-09
### Changed
- Mimetypes are now inferred firstly by file extension, and go's DetectContentType failing that.

### Fixed
- As a result of the mimetype inference change mentioned above, CSS files for example will now be served with the appropriate text/css mimetype.

## [0.3] - 2020-08-16
### Added
- Usage now shows Go's runtime version. Also builds with 0.15.

### Changed
- User Agent strings are now logged.

### Fixed
- Links to filenames with quotes are now html-escaped so they work.

## [0.2] - 2020-06-05
### Added
- Custom bind address with `-b address`.
- Optional TLS with `-c certfile -k keyfile`.

### Changed
- Directory entries are now naturally/alphanumerically sorted.
- Symlinks were made forbidden.
- Sends `Cache-Control: no-store` for HTTP 1.1+ clients that obey it (pretty much all major browsers).
- Rendering performance and size was improved.
- Browsers should not request favicons anymore.

## 0.1 - 2019-09-03
Initial release.


[0.6]: https://github.com/joshuarli/srv/compare/0.6...HEAD
[0.5]: https://github.com/joshuarli/srv/compare/0.4...0.5
[0.4]: https://github.com/joshuarli/srv/compare/0.3...0.4
[0.3]: https://github.com/joshuarli/srv/compare/0.2...0.3
[0.2]: https://github.com/joshuarli/srv/compare/0.1...0.2


bench/main.lua
local threads = {}

function setup(thread)
    table.insert(threads, thread)
    thread:set("tid", table.getn(threads))
end

function init(args)
    responses = {}
end

wrk.method = "GET"
wrk.path = "/"

function response(status, headers, body)
    if responses[status] == nil then
        responses[status] = 1
    else
        responses[status] = responses[status] + 1
    end
end

function done(summary, latency, requests)
    print("wrk is done. response code counts:")
    local freqs = {}
    for _, thread in pairs(threads) do
        for code, freq in pairs(thread:get("responses")) do
            if freqs[code] == nil then
                freqs[code] = freq
            else
                freqs[code] = freqs[code] + freq
            end
        end
    end
    for code, freq in pairs(freqs) do
        print(code, freq)
    end
end


internal/humanize/humanize.go
package humanize

import (
	"fmt"
)

func FileSize(nbytes int64) string {
	if nbytes < 1024 {
		return fmt.Sprintf("%d", nbytes)
	}
	var exp int
	n := float64(nbytes)
	for exp = 0; exp < 4; exp++ {
		n /= 1024
		if n < 1024 {
			break
		}
	}
	return fmt.Sprintf("%.1f%c", float64(n), "KMGT"[exp])
}

func isdigit(b byte) bool { return '0' <= b && b <= '9' }

// Vendored and slightly adapted from github.com/fvbommel/util/blob/master/sortorder/natsort.go
// I wrote my own alphanumeric/natural sort, but it turned out to be inferior to this one.
// Limitations (deliberate tradeoffs for speed and simplicity):
//   - only ASCII digits (0-9) are considered
//   - does not understand signs, scientific notation, or floating point representations
func NaturalLess(str1, str2 string) bool {
	idx1, idx2 := 0, 0
	for idx1 < len(str1) && idx2 < len(str2) {
		c1, c2 := str1[idx1], str2[idx2]
		dig1, dig2 := isdigit(c1), isdigit(c2)
		switch {
		case dig1 != dig2: // Digits before other characters.
			return dig1 // True if LHS is a digit, false if the RHS is one.
		case !dig1: // && !dig2, because dig1 == dig2
			// UTF-8 compares bytewise-lexicographically, no need to decode
			// codepoints.
			if c1 != c2 {
				return c1 < c2
			}
			idx1++
			idx2++
		default: // Digits
			// Eat zeros.
			for ; idx1 < len(str1) && str1[idx1] == '0'; idx1++ {
			}
			for ; idx2 < len(str2) && str2[idx2] == '0'; idx2++ {
			}
			// Eat all digits.
			nonZero1, nonZero2 := idx1, idx2
			for ; idx1 < len(str1) && isdigit(str1[idx1]); idx1++ {
			}
			for ; idx2 < len(str2) && isdigit(str2[idx2]); idx2++ {
			}
			// If lengths of numbers with non-zero prefix differ, the shorter
			// one is less.
			if len1, len2 := idx1-nonZero1, idx2-nonZero2; len1 != len2 {
				return len1 < len2
			}
			// If they're equal, string comparison is correct.
			if nr1, nr2 := str1[nonZero1:idx1], str2[nonZero2:idx2]; nr1 != nr2 {
				return nr1 < nr2
			}
			// Otherwise, the one with less zeros is less.
			// Because everything up to the number is equal, comparing the index
			// after the zeros is sufficient.
			if nonZero1 != nonZero2 {
				return nonZero1 < nonZero2
			}
		}
		// They're identical so far, so continue comparing.
	}
	// So far they are identical. At least one is ended. If the other continues,
	// it sorts last.
	return len(str1) < len(str2)
}


main.go
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"path"
	"runtime"
	"sort"
	"strings"
	"time"

	"github.com/joshuarli/srv/internal/humanize"
)

type context struct {
	srvDir string
}

// We write the shortest browser-valid base64 data string,
// so that the browser does not request the favicon.
const listingPrelude = `<head><link rel=icon href=data:,><style>* { font-family: monospace; } table { border: none; margin: 1rem; } td { padding-right: 2rem; }</style></head>
<table>`

func renderListing(w http.ResponseWriter, r *http.Request, f *os.File) error {
	files, err := f.Readdir(-1)
	if err != nil {
		return err
	}

	io.WriteString(w, listingPrelude)

	sort.Slice(files, func(i, j int) bool {
		// TODO: add switch to make case sensitive
		// TODO: add switch to disable natural sort
		return humanize.NaturalLess(
			strings.ToLower(files[i].Name()),
			strings.ToLower(files[j].Name()),
		)
	})

	var fn, fnEscaped string
	for _, fi := range files {
		fn = fi.Name()
		fnEscaped = url.PathEscape(fn)
		switch m := fi.Mode(); {
		// is a directory - render a link
		case m&os.ModeDir != 0:
			fmt.Fprintf(w, "<tr><td><a href=\"%s/\">%s/</a></td></tr>", fnEscaped, fn)
		// is a regular file - render both a link and a file size
		case m&os.ModeType == 0:
			fs := humanize.FileSize(fi.Size())
			fmt.Fprintf(w, "<tr><td><a href=\"%s\">%s</a></td><td>%s</td></tr>", fnEscaped, fn, fs)
		// otherwise, don't render a clickable link
		default:
			fmt.Fprintf(w, "<tr><td><p style=\"color: #777\">%s</p></td></tr>", fn)
		}
	}

	io.WriteString(w, "</table>")
	return nil
}

func (c *context) handler(w http.ResponseWriter, r *http.Request) {
	// TODO: better log styling
	log.Printf("\t%s [%s]: %s %s %s", r.RemoteAddr, r.UserAgent(), r.Method, r.Proto, r.Host+r.RequestURI)

	// Tell HTTP 1.1+ clients to not cache responses.
	w.Header().Set("Cache-Control", "no-store")

	switch r.Method {
	case http.MethodGet:
		// Filenames could contain special uri characters, so we use r.RequestURI
		// instead of r.URL.Path.
		// XXX: Might also have to do QueryUnescape (and then also QueryEscape in the renderer),
		// but haven't run into that as a need in my usage.
		fp, err := url.PathUnescape(r.RequestURI)
		if err != nil {
			http.Error(w, fmt.Sprintf("failed to path unescape: %s", err), http.StatusInternalServerError)
			return
		}
		fp = path.Join(c.srvDir, fp)
		fi, err := os.Lstat(fp)
		if err != nil {
			// NOTE: errors.Is is generally preferred, since it can unwrap errors created like so:
			//     fmt.Errorf("can't read file: %w", err)
			// But in this case we just want to check right after a stat.
			if os.IsNotExist(err) {
				http.Error(w, "file not found", http.StatusNotFound)
				return
			}
			http.Error(w, fmt.Sprintf("failed to stat file: %s", err), http.StatusInternalServerError)
			return
		}

		f, err := os.Open(fp)
		if err != nil {
			http.Error(w, fmt.Sprintf("failed to open file: %s", err), http.StatusInternalServerError)
			return
		}
		defer f.Close()

		switch m := fi.Mode(); {
		// is a directory - serve an index.html if it exists, otherwise generate and serve a directory listing
		case m&os.ModeDir != 0:
			// XXX: if a symlink has name "index.html", it will be served here.
			// i could add an extra lstat here, but the scenario is just too rare
			// to justify the additional file operation.
			html, err := os.Open(path.Join(fp, "index.html"))
			if err == nil {
				io.Copy(w, html)
				html.Close()
				return
			}
			html.Close()
			err = renderListing(w, r, f)
			if err != nil {
				http.Error(w, "failed to render directory listing: "+err.Error(), http.StatusInternalServerError)
			}
		// is a regular file - serve its contents
		case m&os.ModeType == 0:
			// This deduces a mimetype from the file extension first, then falls back to DetectContentType.
			// io.Copy'ing would only DetectContentType, which is insufficient for like, css files.
			http.ServeContent(w, r, fp, time.Time{}, f)
		// is a symlink - refuse to serve
		case m&os.ModeSymlink != 0:
			// TODO: add a flag to allow serving symlinks
			http.Error(w, "file is a symlink", http.StatusForbidden)
		default:
			http.Error(w, "file isn't a regular file or directory", http.StatusForbidden)
		}
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

func die(format string, v ...interface{}) {
	fmt.Fprintf(os.Stderr, format, v...)
	os.Stderr.Write([]byte("\n"))
	os.Exit(1)
}

// VERSION passed at build time
var VERSION = "unknown"

func main() {
	flag.Usage = func() {
		die(`srv %s (go version %s)

usage: %s [-q] [-p port] [-c certfile -k keyfile] directory

directory       path to directory to serve (default: .)

-q              quiet; disable all logging
-p port         port to listen on (default: 8000)
-b address      listener socket's bind address (default: 127.0.0.1)
-c certfile     optional path to a PEM-format X.509 certificate
-k keyfile      optional path to a PEM-format X.509 key
`, VERSION, runtime.Version(), os.Args[0])
	}

	var quiet bool
	var port, bindAddr, certFile, keyFile string
	flag.BoolVar(&quiet, "q", false, "")
	flag.StringVar(&port, "p", "8000", "")
	flag.StringVar(&bindAddr, "b", "127.0.0.1", "")
	flag.StringVar(&certFile, "c", "", "")
	flag.StringVar(&keyFile, "k", "", "")
	flag.Parse()

	certFileSpecified := certFile != ""
	keyFileSpecified := keyFile != ""
	if certFileSpecified != keyFileSpecified {
		die("You must specify both -c certfile -k keyfile.")
	}

	listenAddr := net.JoinHostPort(bindAddr, port)
	_, err := net.ResolveTCPAddr("tcp", listenAddr)
	if err != nil {
		die("Could not resolve the address to listen to: %s", listenAddr)
	}

	srvDir := "."
	posArgs := flag.Args()
	if len(posArgs) > 0 {
		srvDir = posArgs[0]
	}
	f, err := os.Open(srvDir)
	if err != nil {
		die(err.Error())
	}
	defer f.Close()
	if fi, err := f.Stat(); err != nil || !fi.IsDir() {
		die("%s isn't a directory.", srvDir)
	}

	c := &context{
		srvDir: srvDir,
	}

	if quiet {
		log.SetFlags(0) // disable log formatting to save cpu
		log.SetOutput(io.Discard)
	}

	http.HandleFunc("/", c.handler)

	if certFileSpecified && keyFileSpecified {
		log.Printf("\tServing %s over HTTPS on %s", srvDir, listenAddr)
		err = http.ListenAndServeTLS(listenAddr, certFile, keyFile, nil)
	} else {
		log.Printf("\tServing %s over HTTP on %s", srvDir, listenAddr)
		err = http.ListenAndServe(listenAddr, nil)
	}

	die(err.Error())
}
